### ISSUE TITLE
stack 1.1.2 and docker do not work on OS X

### ISSUE THREAD

**Issue by vizowl** [commercialhaskell/stack#2182]
After upgrading to stack-1.1.2, projects with docker enabled no longer work on a mac. `docker build` fails with with error `SQLite3 returned ErrorCan'tOpen while attempting to perform open "~/.stack/docker.db".`

I deleted the docker.db file and still got the same error. There is no problem with stack 1.1.0.

@christiantakle reported in #2157 that this problem occurs with both docker-toolbox and docker beta.

**Comment by borsboom** [comment]
How are you running Docker on Mac?  Are you using the closed beta Docker for Mac, or are you using Docker Toolkit, or are you using https://github.com/borsboom/vagrant-stack-docker?  I'm using the latter, and it's working for me with stack-1.1.2.

Keep in mind that Stack's Docker integration only officially supports running on Linux hosts, so it's entirely YMMV if you try to use it on OS X.  That said, I have high hopes for official OS X support once Docker for Mac is publicly released.

**Comment by borsboom** [comment]
Oh, also: how did you install stack-1.1.2?

**Comment by vizowl** [comment]
@borsboom I am running the closed private beta - so I naturally don't expect everything to work :) But I figured it was useful to know that it wasn't working.

I installed stack via homebrew.

**Comment by christiantakle** [comment]
The fault seems to be with the homebrew version. I've tried the following:

* Docker Toolbox + stack v1.1.0 (nix) => Ok
* Docker for Mac + stack v1.1.0 (nix) => Unknown
* Docker Toolbox + stack v1.1.2 (`stack upgrade` from the nix install) => Ok
* Docker for Mac + stack v1.1.2 (`stack upgrade` from the nix install) => Unknown
* Docker for Mac + stack v1.1.2 (homebrew) => Fault
* Docker Toolbox + stack v1.1.2 (homebrew) => Fault

**Comment by vizowl** [comment]
I can confirm the problem is with the homebrew build of stack - I just downloaded stack 1.1.2 from https://www.stackage.org/stack/osx-x86_64 and I do not observe this problem. This looks like more of a homebrew issue - but I see that @borsboom provides the homebrew formula ... so would you like me to close this issue and file a new one in homebrew?

**Comment by borsboom** [comment]
I do patch the Homebrew formula when new versions of Stack come out, but I didn't originate it.  As I understand it, the way Homebrew builds is by grabbing the source tarball from Github and then using cabal-install (and its solver) to build it.  Unfortunately this means they may use different versions of the dependencies than we use for the official binaries and packages, and I suspect that's why we're seeing different behaviour.  This is exacerbated by the fact that they don't want to get the sdist from hackage (which has tighter bounds due to using `stack upload --pvp-bounds=both`) but instead use the source directly from Github (which doesn't have very many bounds set, as we use Stackage).  Perhaps the thing to do is upload the sdist with pvp-bounds to the Github release as well and have Homebrew use that instead of the source tarball generated directly from the repo.  I've been tempted to include a `cabal.config` with the source code to force them to use exactly the right versions of dependencies (since even --pvp-bounds=both are a bit looser than would be ideal), but it's extra work and also will cause other problems: in particular, Homebrew is switching to ghc-8.0.1 already because they can't support multiple GHC versions, and a lot of the pinned package versions won't work with ghc-8.0.

Anyway, all this is to say that I honestly don't know the best way to proceed.  Maybe you can open issue against in Homebrew so we can get their input and hash it out.

**Comment by borsboom** [comment]
One thing to add: we're [working on a simpler way to install Stack using official bindists](https://github.com/haskell-lang/haskell-lang/issues/42) and once that's in place I'm thinking we should stop recommending Homebrew as a way to install Stack (although I think Stack should stay in Homebrew regardless).

**Comment by borsboom** [comment]
Upon further reflection, the ideal way to do this is probably to have the formula use cabal-install to bootstrap Stack, but then use the cabal-install-built `stack` to rebuild Stack properly.

inviting @ilovezfs: do you have any thoughts about this?

**Comment by ilovezfs** [comment]
>but then use the cabal-install-built stack to rebuild Stack properly.

Yeah bootstrapping sounds like an interesting idea, assuming the final product is what brew is then installing (and, therefore, what's in the bottles) as opposed to something the user is supposed to build after-the-fact. Can you describe how that would work? Can the initially built version be used to build the stage 2 version within the same (or another) cabal sandbox? Does it have to gather additional resources to build the stage 2 version? Would a stage 3 version be needed?

**Comment by borsboom** [comment]
Right, the final product is what brew would install.  This is how we bootstrap Stack on new platforms ourselves.  There would be two stages, and since Stack uses completely separate sandboxes from cabal-install, the second stage would automatically be built in a different sandbox.  So the basic process goes like this (the details would, I'm sure, vary in the Homebrew case)

```
git clone git@github.com/commercialhaskell/stack.git
cd stack
cabal sandbox init
cabal install
.cabal-sandbox/bin/stack setup
.cabal-sandbox/bin/stack --local-bin-path=/destination/path install
```

Now the final executable will be in `/destination/path/stack`.

**Comment by borsboom** [comment]
Oh and yes, `stack` has to download Stackage snapshots and packages to do its work.

**Comment by ilovezfs** [comment]
Does `.cabal-sandbox/bin/stack setup` stay within the buildpath? If not, where else does it try to write? Is that configurable?

>git clone git@github.com/commercialhaskell/stack.git

As long as there's a specific tag+revision this can work from, it's fine: e.g., https://github.com/Homebrew/homebrew-core/blob/master/Formula/osquery.rb#L5-L7

The same can be used in resource blocks.

>Oh and yes, stack has to download Stackage snapshots and packages to do its work.

Would we know the exact versions in advance so they can go in resource blocks with checksums and be made available within the buildpath?

**Comment by borsboom** [comment]
> Does .cabal-sandbox/bin/stack setup stay within the buildpath? If not, where else does it try to write? Is that configurable?

It can, by setting the STACK_ROOT environment variable to be within the buildpath.

> git clone git@github.com/commercialhaskell/stack.git

The details of where the source codes come from don't really matter.  It can grab the source tarball from Github just like the current formula does.

> Would we know the exact versions in advance so they can go in resource blocks with checksums and be made available within the buildpath?

That's harder... but do you do this for cabal-install based installs?  Because what Stack does is similar to cabal-install downloading a package index and packages from Hackage.

**Comment by ilovezfs** [comment]
>do you do this for cabal-install based installs?

Nope. And it's not an absolute requirement, just is the aspect of what you're describing that will get the closest scrutiny due to security concerns. Usually if the backend is using https+$SomeKindOfChecksumming it's OK.

**Comment by ilovezfs** [comment]
>Because what Stack does is similar to cabal-install downloading a package index and packages from Hackage.

If there's anything that describes the implementation details in words, that will help make the case.

**Comment by borsboom** [comment]
Hmm, I'm not aware of a current document that describes all the implementation details.  [This blog post](https://www.fpcomplete.com/blog/2015/07/package-security-in-stack) does describe the approach to security (and, since then, we've also implemented [author package signing](https://www.fpcomplete.com/blog/2016/05/stack-security-gnupg-keys)).  In a nutshell, Stack does use HTTPS everywhere and verifies checksums and GPG signatures (one of Stack's motivations was to have better security than cabal-install).

**Comment by ilovezfs** [comment]
Yeah this is definitely sounding like a promising way to go.

**Comment by ilovezfs** [comment]
@borsboom OK, looks I got this working using your instructions above if you'd like to try it out.

```
(buildpath/"cabal.config").write("allow-newer: base,transformers\n")
mkdir buildpath/"stack_root"
ENV["STACK_ROOT"] = buildpath/"stack_root"
cabal_sandbox do
  system "cabal", "install"
  system "./.cabal-sandbox/bin/stack", "setup"
  system "./.cabal-sandbox/bin/stack", "--local-bin-path=#{bin}", "install"
end
```

**Comment by ilovezfs** [comment]
@borsboom FYI https://github.com/Homebrew/homebrew-core/pull/1480

**Comment by borsboom** [comment]
This bug appears to be introduced with persistent-sqlite-2.5.0.1 (see https://github.com/yesodweb/persistent/pull/561#issuecomment-222329087).  I've added a version constraint to prevent using that version.

**Comment by borsboom** [comment]
Fixed the bug in upstream persistent-sqlite-2.5.0.1 as well: https://github.com/yesodweb/persistent/pull/565
